### PR TITLE
Add a mode to next build to make it easier to debug memory issues

### PR DESCRIPTION
**Pull request by mknichel**
This PR adds a `--experimental-debug-memory-usage` flag to `next build` to make it easier to debug memory performance. This mode does the following things:

- Periodically prints the current memory usage of the process
- Records garbage collection events and warns about long running GC events
- Kills the process if it detects GC thrashing near heap limit
- Automatically takes a heap snapshot if heap usage rises above 70% of the total heap
- Automatically takes a heap snapshot if the process is close to running out of memory
- Prints a report at the end of the build with information about peak memory usage and time spent in garbage collection

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: eba8a4a6b9bea11746c7c9a1fa1a310c44594101

`pnpm test test/integration/amphtml/test/index.test.js`
- AMP Usage > production mode > should have amp optimizer in trace
- AMP Usage > production mode > With AMP context > should render the normal page that uses the AMP hook
- AMP Usage > production mode > With AMP context > should render the AMP page that uses the AMP hook
- AMP Usage > production mode > With AMP context > should render nested normal page with AMP hook
- AMP Usage > production mode > With AMP context > should render nested AMP page with AMP hook
- AMP Usage > production mode > With basic AMP usage > should render the page as valid AMP
- AMP Usage > production mode > With basic AMP usage > should render the page without leaving render target
- AMP Usage > production mode > With basic AMP usage > should not output client pages for AMP only
- AMP Usage > production mode > With basic AMP usage > should not output client pages for AMP only with config exported after declaration
- AMP Usage > production mode > With basic AMP usage > should drop custom scripts
- AMP Usage > production mode > With basic AMP usage > should not drop custom amp scripts
- AMP Usage > production mode > With basic AMP usage > should optimize clean
- AMP Usage > production mode > With basic AMP usage > should auto import extensions
- AMP Usage > production mode > With basic usage > should render the page
- AMP Usage > production mode > canonical amphtml > should render link rel amphtml
- AMP Usage > production mode > canonical amphtml > should render amphtml from provided rel link
- AMP Usage > production mode > canonical amphtml > should render link rel amphtml with existing query
- AMP Usage > production mode > canonical amphtml > should render the AMP page that uses the AMP hook
- AMP Usage > production mode > canonical amphtml > should render a canonical regardless of amp-only status (explicit)
- AMP Usage > production mode > canonical amphtml > should not render amphtml link tag with no AMP page
- AMP Usage > production mode > canonical amphtml > should remove conflicting amp tags
- AMP Usage > production mode > canonical amphtml > should allow manually setting canonical
- AMP Usage > production mode > canonical amphtml > should allow manually setting amphtml rel
- AMP Usage > production mode > combined styles > should combine style tags
- AMP Usage > production mode > combined styles > should remove sourceMaps from styles

<details>
<summary>Expand output</summary>

  ● AMP Usage › production mode › should have amp optimizer in trace

    ENOENT: no such file or directory, open '/root/actions-runner/_work/next.js/next.js/test/integration/amphtml/.next/next-server.js.nft.json'

      56 |       it('should have amp optimizer in trace', async () => {
      57 |         const trace = JSON.parse(
    > 58 |           readFileSync(join(appDir, '.next/next-server.js.nft.json'), 'utf8')
         |                       ^
      59 |         )
      60 |         expect(
      61 |           trace.files.some((file) =>

      at Object.<anonymous> (integration/amphtml/test/index.test.js:58:23)

  ● AMP Usage › production mode › With basic usage › should render the page

    FetchError: request to http://localhost:40559/ failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › With basic AMP usage › should render the page as valid AMP

    FetchError: request to http://localhost:40559/?amp=1 failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › With basic AMP usage › should render the page without leaving render target

    FetchError: request to http://localhost:40559/special-chars failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › With basic AMP usage › should not output client pages for AMP only

    page.goto: net::ERR_CONNECTION_REFUSED at http://localhost:40559/nav
    Call log:
      - navigating to "http://localhost:40559/nav", waiting until "load"

      274 |     opts?.beforePageLoad?.(page)
      275 |
    > 276 |     await page.goto(url, { waitUntil: 'load' })
          |                ^
      277 |   }
      278 |
      279 |   back(options): BrowserInterface {

      at BrowserInterface.goto (lib/browsers/playwright.ts:276:16)
      at webdriver (lib/next-webdriver.ts:129:3)
      at Object.<anonymous> (integration/amphtml/test/index.test.js:96:27)

  ● AMP Usage › production mode › With basic AMP usage › should not output client pages for AMP only with config exported after declaration

    page.goto: net::ERR_CONNECTION_REFUSED at http://localhost:40559/nav
    Call log:
      - navigating to "http://localhost:40559/nav", waiting until "load"

      274 |     opts?.beforePageLoad?.(page)
      275 |
    > 276 |     await page.goto(url, { waitUntil: 'load' })
          |                ^
      277 |   }
      278 |
      279 |   back(options): BrowserInterface {

      at BrowserInterface.goto (lib/browsers/playwright.ts:276:16)
      at webdriver (lib/next-webdriver.ts:129:3)
      at Object.<anonymous> (integration/amphtml/test/index.test.js:105:27)

  ● AMP Usage › production mode › With basic AMP usage › should drop custom scripts

    FetchError: request to http://localhost:40559/custom-scripts failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › With basic AMP usage › should not drop custom amp scripts

    FetchError: request to http://localhost:40559/amp-script?amp=1 failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › With basic AMP usage › should optimize clean

    FetchError: request to http://localhost:40559/only-amp failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › With basic AMP usage › should auto import extensions

    FetchError: request to http://localhost:40559/auto-import failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › With AMP context › should render the normal page that uses the AMP hook

    FetchError: request to http://localhost:40559/use-amp-hook failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › With AMP context › should render the AMP page that uses the AMP hook

    FetchError: request to http://localhost:40559/use-amp-hook?amp=1 failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › With AMP context › should render nested normal page with AMP hook

    FetchError: request to http://localhost:40559/nested failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › With AMP context › should render nested AMP page with AMP hook

    FetchError: request to http://localhost:40559/nested?amp=1 failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › canonical amphtml › should render link rel amphtml

    FetchError: request to http://localhost:40559/use-amp-hook failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › canonical amphtml › should render amphtml from provided rel link

    FetchError: request to http://localhost:40559/use-amp-hook.amp failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › canonical amphtml › should render link rel amphtml with existing query

    FetchError: request to http://localhost:40559/use-amp-hook?hello=1 failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › canonical amphtml › should render the AMP page that uses the AMP hook

    FetchError: request to http://localhost:40559/use-amp-hook?amp=1 failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › canonical amphtml › should render a canonical regardless of amp-only status (explicit)

    FetchError: request to http://localhost:40559/only-amp failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › canonical amphtml › should not render amphtml link tag with no AMP page

    FetchError: request to http://localhost:40559/normal failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › canonical amphtml › should remove conflicting amp tags

    FetchError: request to http://localhost:40559/conflicting-tag?amp=1 failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › canonical amphtml › should allow manually setting canonical

    FetchError: request to http://localhost:40559/manual-rels?amp=1 failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › canonical amphtml › should allow manually setting amphtml rel

    FetchError: request to http://localhost:40559/manual-rels failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › combined styles › should combine style tags

    FetchError: request to http://localhost:40559/styled?amp=1 failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● AMP Usage › production mode › combined styles › should remove sourceMaps from styles

    FetchError: request to http://localhost:40559/styled?amp=1 failed, reason: connect ECONNREFUSED 127.0.0.1:40559

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js mknichel/memory-debugging-mode | Change |
| - | - | - | - |
| buildDuration | 14s | 14s | N/A |
| buildDurationCached | 7.6s | 6.5s | N/A |
| nodeModulesSize | 199 MB | 199 MB | ⚠️ +65.4 kB |
| nextStartRea..uration (ms) | 403ms | 399ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js mknichel/memory-debugging-mode | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 31.4 kB | 31.4 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | ✓ |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 242 B | 242 B | ✓ |
| main-HASH.js gzip | 32.2 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.68 kB | 1.68 kB | N/A |
| Overall change | 99.3 kB | 99.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js mknichel/memory-debugging-mode | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js mknichel/memory-debugging-mode | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.21 kB | 4.21 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.57 kB | 6.57 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js mknichel/memory-debugging-mode | Change |
| - | - | - | - |
| _buildManifest.js gzip | 481 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js mknichel/memory-debugging-mode | Change |
| - | - | - | - |
| index.html gzip | 530 B | 529 B | N/A |
| link.html gzip | 542 B | 542 B | ✓ |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 542 B | 542 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js mknichel/memory-debugging-mode | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95.3 kB | 95.3 kB | N/A |
| page.js gzip | 3.04 kB | 3.04 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js mknichel/memory-debugging-mode | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 626 B | 624 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 990 B | 990 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js mknichel/memory-debugging-mode | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 170 kB | 170 kB | ✓ |
| app-page-exp..prod.js gzip | 97 kB | 97 kB | ✓ |
| app-page-tur..prod.js gzip | 98.8 kB | 98.8 kB | ✓ |
| app-page-tur..prod.js gzip | 93 kB | 93 kB | ✓ |
| app-page.run...dev.js gzip | 144 kB | 144 kB | ✓ |
| app-page.run..prod.js gzip | 91.5 kB | 91.5 kB | ✓ |
| app-route-ex...dev.js gzip | 21.4 kB | 21.4 kB | ✓ |
| app-route-ex..prod.js gzip | 15.2 kB | 15.2 kB | ✓ |
| app-route-tu..prod.js gzip | 15.2 kB | 15.2 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route.ru...dev.js gzip | 21.1 kB | 21.1 kB | ✓ |
| app-route.ru..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.82 kB | 9.82 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-turbo...prod.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim...dev.js gzip | 23.1 kB | 23.1 kB | ✓ |
| pages.runtim..prod.js gzip | 22.4 kB | 22.4 kB | ✓ |
| server.runti..prod.js gzip | 51 kB | 51 kB | ✓ |
| Overall change | 945 kB | 945 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js mknichel/memory-debugging-mode | Change |
| - | - | - | - |
| 0.pack gzip | 1.58 MB | 1.58 MB | ⚠️ +753 B |
| index.pack gzip | 106 kB | 106 kB | ⚠️ +137 B |
| Overall change | 1.69 MB | 1.69 MB | ⚠️ +890 B |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: eba8a4a6b9bea11746c7c9a1fa1a310c44594101

File: packages/next/src/bin/next.ts [hunk 1]
```diff
@@ -105,6 +105,10 @@ program
     )}`
   )
   .option('-d, --debug', 'Enables a more verbose build output.')
+  .option(
+    '--debug-memory-usage',
```
Comment (ijjk) [inline packages/next/src/bin/next.ts:109]:
```suggestion
    '--experimental-debug-memory-usage',
```
We might wanna start with this being experimental while we iterate on it

Comment (mknichel) [reply packages/next/src/bin/next.ts]: Done.

File: packages/next/src/build/index.ts [hunk 2]
```diff
@@ -1626,6 +1629,7 @@ export default async function build(
             'edge-server',
           ]).then((res) => {
             durationInSeconds += res.duration
+            traceMemoryUsage('Finished edge-server compilation', nextBuildSpan)
```
Comment (ijjk) [inline packages/next/src/build/index.ts:1632]: Do we want to grab the memory usage from inside of the worker when using the `webpackBuildWorker` mode?

**Review by ijjk — COMMENTED** [review]
A small test to ensure the flag is properly honored would be solid to avoid regressing

**Review by ijjk — APPROVED** [review]

**Comment by danielhjacobs** [comment]
After this PR, https://nextjs.org/ no longer works with PerformanceObserver undefined. See #64211.

**Comment by danielhjacobs** [comment]
https://github.com/vercel/next.js/blob/canary/packages/next/src/lib/memory/gc-observer.ts should be changed, see also #30387.

**Comment by danielhjacobs** [comment]
Suggestion:

```ts
const obs = window.PerformanceObserver ? new PerformanceObserver((list) => {
  const entry = list.getEntries()[0]
  gcEvents.push(entry)

  if (entry.duration > LONG_RUNNING_GC_THRESHOLD_MS) {
    warn(bold(`Long running GC detected: ${entry.duration.toFixed(2)}ms`))
  }
}) : undefined;

export function startObservingGc() {
  obs?.observe({ entryTypes: ['gc'] })
}

export function stopObservingGc() {
  obs?.disconnect()
}
```

**Comment by ijjk** [comment]
@danielhjacobs can you provide more information why you think this PR is related, this is only changing code that runs in the node environment and doesn't affect client side code.

**Pull request merged**